### DISCUSSION TITLE
Convention

### DISCUSSION THREAD

**Discussion by vaskes79** [Shadow-Shamans/first-race#15] [Announcements]
## Contention

Документ описывает структуру проекта и правила разработки.

### Файловая структура проекта.

```
.
├── app
│   ├── App.less
│   ├── App.test.tsx
│   ├── App.tsx
│   ├── index.ts
│   └── themes
│       ├── dark.ts
│       ├── index.ts
│       └── light.ts
├── assets
│   └── logo.svg
├── shared
│   └── Button
│       ├── Button.less
│       ├── Button.test.tsx
│       ├── Button.tsx
│       └── index.ts
├── feature
│   └── Auth
│       ├── authSlice.ts
│       ├── Auth.test.tsx
│       ├── Auth.tsx
│       ├── constants.ts
│       ├── index.ts
│       └── types.ts
├── game
│   ├── assets
│   │   ├── sounds
│   │   │   ├── music
│   │   │   │   └── base-theme.wav
│   │   │   └── train.wav
│   │   └── sprites
│   │       ├── station.png
│   │       └── train.png
│   └── Game.ts
├── main.tsx
├── pages
│   ├── ForumPage.tsx
│   ├── HomePage.tsx
│   ├── index.ts
│   ├── LoginPage.tsx
│   ├── PrivateRouteComponent.tsx
│   └── router.ts
├── store
│   ├── index.ts
│   ├── RootReducer.ts
│   └── store.ts
└── vite-env.d.ts

```

### Последовательность выполнения работ.

1. [#1](https://github.com/Shadow-Shamans/first-race/issues/1) инициализация проекта

- установить [ant](https://ant.design/docs/react/getting-started)
- настроить [переключение темы в анте](https://betterprogramming.pub/how-to-toggle-dark-theme-with-ant-design-5-0-eb68552f62b8)
- установить [React Router](https://reactrouter.com/en/main/start/tutorial)
- установить настроить [RTK](https://redux-toolkit.js.org/tutorials/typescript)
- создать [слайс User](https://redux-toolkit.js.org/tutorials/typescript#application-usage) для хранения данных пользователя
- настроить [RTKQuery](https://redux-toolkit.js.org/tutorials/rtk-query)
- установить [ReduxPersist](https://blog.logrocket.com/persist-state-redux-persist-redux-toolkit-react/)
- добавить в white list persist store Auth для хранения позльзовательских настроект между сессиями
- опционально подключить [Linaria](https://github.com/callstack/linaria#installation)

2. [#4](https://github.com/Shadow-Shamans/first-race/issues/4) React Router и React Router DOM

Для каждой страницы нужно создать роут и отдельный драфт компонент.

- Сделать компонент приватного роута.
- Роут который редиректит на 404 при запросе несуществющего
- Главная страница — мини-лендинг с презентацией игры. Сейчас можно сделать сразу вход или авторизацию.
- Страница входа.
- Страница регистрации.
- Страница пользователя — страница с информацией о пользователе, возможностью загрузить и сменить аватар, изменить пароль (используйте [API](https://ya-praktikum.tech/api/v2/swagger/#/)).
- Страница лидерборда — страница с таблицей всех игроков, в которой они отсортированы от наибольшего числа баллов к наименьшему. Сейчас использовать API необязательно, достаточно продумать экран и сверстать его на React.
- Авторизация — регистрация и страница входа (уже с API, берите API из предыдущего модуля).
- Форум — страница, на которой пользователи могут создавать темы, оставлять в этих темах комментарии и отвечать на них. API вам предстоит написать самостоятельно в восьмом спринте.

Эти два этап должны идти первыми, они закладывают фундамент для дальнейшей неблокируюещей работы. На время выполнения этих этапов команда может заниматься, часть заблокированной команды может заниматься теорией. Предоположительное время работ 2 дня. Дальнейшие работы могут быть в произвольном порядке.

### gitflow

- создаем ветку `dev` от `main`
- для работы над задачей создаем ветку `4-react-router` от ветки `dev` делаем таску
- перед отпрвкой в репозиторий делаем `git fetch` затем rebase на`dev` после этого отправляем в remote rep
- в репозитории делаем mr с номером таски #4 копируем пр и в чат кидаем на ревью
- на каждый коммент ревью делаем отдельный коммит, если считаете, что комменты однотипные, можно объединить в один коммит
- тот кто ревьювил резолвит коммент после правок

---

**Comment by dahah-babah** [comment]
В целом супер, но я бы добавила:

**В 1 этап:**

Нужно добавить/подогнать конфиги под себя, обсудить какие правила хотим:

+ настройка tsconfig
+ настройка eslint
+ настройка prettier
+ настройка stylelint
+ настройка editorconfig
+ настройка husky/precommit

Если решили брать линарию - давайте тоже подключим ее в этом этапе и уберем `less` файлы из структуры, вместо этого давайте использовать `styles.ts`

Ссылки по линарии продублирую:

- Нстройка с vite https://github.com/callstack/linaria/blob/master/docs/BUNDLERS_INTEGRATION.md#vite
- Темизация https://github.com/callstack/linaria/blob/master/docs/THEMING.md

**По гитфлоу:**

- Предлагаю в названии ветки разделять номер и название через слеш `4/react-router` (мне просто так кажется нагляднее) + можем добавить тип задачи (feature/fix)
- По описанию к мр-у есть предложение добавить чек лист, например описание, ссылку на задачу и тп
- Не вливаем в dev-ветку без апрува (один минимум)

**Reply by vaskes79** [reply]
@dahah-babah 

- eslint
- tsconfig
- prettier
- husky (lefthook) 

У нас уже есть в проекте.

**Reply by vaskes79** [reply]
@dahah-babah если мы решили использовать Ant, я бы не стал тратить время на настройку и отладку работы styled-components. less в анте идет из коробки ничего не нужно настраивать, почти. Если ты хочешь попробуй подключить и настроить linary вместе с антом и создай ветку. Если получится и всех устроит вмержим и будем c ней работать.

**Reply by vaskes79** [reply]
по именованию веток согласен, еднственное номер, лучше всегда в начале ставить, проще работать с автокомплитом бранчей в консоле

**Reply by vaskes79** [reply]
`По описанию к мр-у есть предложение добавить чек лист, например описание, ссылку на задачу и тп`

можно попробовать, но мне кажется будет лениво, это делать, а вообще идея хорошая.

**Reply by dahah-babah** [reply]
1. Базовые конфиги есть, но скорее всего мы их будем изменять/дорабатывать по себя
2. Не вижу проблемы подключить линарию сразу, но можем попробовать сделать это отдельно, не вопрос)
3. Я вроде номер и поставила в начале `4/react-router`
4. Мы можем использовать шаблон для описания, но если не хотим заморачиваться, то просто будем помнить, что хорошая практика - кратко описать что сделано

**Comment by dahah-babah** [comment]
**Еще момент по структуре:**

Давайте разделим ассеты на icons (svg), images (png) и fonts, чтобы не было мешанины

**Reply by vaskes79** [reply]
для icons в ант есть отдельный компонент, деление по папкам в assets кажется избыточным, мы все равно туда редко будем заглядывать, она нужна будет в основном для deploya для раздачи статики.

**Reply by dahah-babah** [reply]
Не думаю, что прям все иконки буду из ант-а. Даже если заглядываешь редко - порядок важен
